plugin/src/main/java/com/playmonumenta/scriptedquests/zones/zone/Zone.java

```diff
@@ -0,0 +1,273 @@
+		return splitByOverlap(overlap, null);
+	}
+
+	/*
+	 * Split all fragments of this zone by an overlapping zone,
+	 * marking otherZone as the parent of the exact overlap fragment if
+	 * it exists. Otherwise, the exact overlap fragment is discarded.
+	 *
+	 * Returns true if the zone being overlapped has been completely
+	 * eclipsed by the other zone.
+	 */
+	public boolean splitByOverlap(BaseZone overlap, Zone otherZone) {
+		ArrayList<ZoneFragment> newFragments = new ArrayList<ZoneFragment>();
+		for (ZoneFragment fragment : mFragments) {
+			BaseZone subOverlap = fragment.overlappingZone(overlap);
+
+			if (subOverlap == null)
+			{
+				newFragments.add(fragment);
+				continue;
+			}
+
+			newFragments.addAll(fragment.splitByOverlap(subOverlap, otherZone));
+			fragment.invalidate();
+		}
+		mFragments = newFragments;
+		return newFragments.size() == 0;
+	}
+
+	/*
+	 * Minimize the number of uneclipsed fragments.
+	 *
+	 * This works with only one zone's fragments at a time, and doesn't
+	 * need to be run again. This reduces n significantly for runtime.
+	 *
+	 * TODO Implement this based on the same function in:
+	 * https://github.com/NickNackGus/monumenta-zone-prototype/blob/master/python/lib/zone/zone.py
+	 * Also needs to invalidate fragments of merged zones.
+	 */
+	/*
+	public void defragment() {
+		if (mFragments.size() < 2) {
+			return;
+		}
+
+		HashMap<int, HashMap<LinkedHashSet<int>, ZoneFragment>> allMergedCombinations = new HashMap<int, HashMap<LinkedHashSet<int>, ZoneFragment>>;
+		private ArrayList<ZoneFragment> recursiveOptimalDefrag(HashMap<int, HashMap<LinkedHashSet<int>, ZoneFragment>> allMergedCombinations,
+		                                                       ArrayList<ZoneFragment> resultsSoFar,
+		                                                       LinkedHashSet<int> remainingIds);
+
+		return;
+	}
+	*/
+
+	public ZoneLayer getLayer() {
+		return mLayer;
+	}
+
+	public String getLayerName() {
+		return mLayer.getName();
+	}
+
+	public String getName() {
+		return mName;
+	}
+
+	public ArrayList<ZoneFragment> getZoneFragments() {
+		ArrayList<ZoneFragment> result = new ArrayList<ZoneFragment>();
+		result.addAll(mFragments);
+		return result;
+	}
+
+	public LinkedHashSet<String> getProperties() {
+		LinkedHashSet<String> result = new LinkedHashSet<String>();
+		result.addAll(mProperties);
+		return result;
+	}
+
+	public boolean hasProperty(String propertyName) {
+		return mProperties.contains(propertyName);
+	}
+
+	private static void applyProperty(HashMap<String, ArrayList<String>> propertyGroups, LinkedHashSet<String> currentProperties, String propertyName) throws Exception {
+		applyProperty(propertyGroups, currentProperties, propertyName, false);
+	}
+
+	private static void applyProperty(HashMap<String, ArrayList<String>> propertyGroups, LinkedHashSet<String> currentProperties, String propertyName, boolean remove) throws Exception {
+		if (propertyName == null) {
```

Comment:
Maybe also check for isEmpty() here? Since otherwise charAt(0) will error.

Reply:
Added a commit for this.